### PR TITLE
Fix resolving requirements with percent encoded characters

### LINKED ISSUE #143: Resolution fails due to encoding issue with version strings containing a `+`
The following `requirements.txt` can be resolved with `pip`, but not with `python-inspector`.

```
--extra-index-url https://download.pytorch.org/whl/cpu
torch==2.0.0+cpu
```

Outcome:
```
python-inspector -p 3.10  --operating-system linux --json-pdt x.json --analyze-setup-py-insecurely --requirement requirements.txt --verbose
Resolving dependencies...
direct_dependencies:
 DependentPackage(purl='pkg:pypi/torch@2.0.0%2Bcpu', extracted_requirement='torch==2.0.0+cpu', scope='install')
environment: Environment(python_version='310', operating_system='linux')
repos:
 PypiSimpleRepository(index_url='https://pypi.org/simple', credentials=None)
Traceback (most recent call last):
  File "/home/frank/.local/lib/python3.10/site-packages/resolvelib/resolvers.py", line 397, in resolve
    self._add_to_criteria(self.state.criteria, r, parent=None)
  File "/home/frank/.local/lib/python3.10/site-packages/resolvelib/resolvers.py", line 174, in _add_to_criteria
    raise RequirementsConflicted(criterion)
resolvelib.resolvers.RequirementsConflicted: Requirements conflict: <Requirement('torch==2.0.0+cpu')>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/frank/.local/lib/python3.10/site-packages/python_inspector/resolve_cli.py", line 247, in resolve_dependencies
    resolution_result: Dict = resolver_api(
  File "/home/frank/.local/lib/python3.10/site-packages/python_inspector/api.py", line 263, in resolve_dependencies
    resolution, purls = resolve(
  File "/home/frank/.local/lib/python3.10/site-packages/python_inspector/api.py", line 322, in resolve
    resolved_dependencies, packages = get_resolved_dependencies(
  File "/home/frank/.local/lib/python3.10/site-packages/python_inspector/api.py", line 360, in get_resolved_dependencies
    resolver_results = resolver.resolve(requirements=requirements, max_rounds=max_rounds)
  File "/home/frank/.local/lib/python3.10/site-packages/resolvelib/resolvers.py", line 546, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/home/frank/.local/lib/python3.10/site-packages/resolvelib/resolvers.py", line 399, in resolve
    raise ResolutionImpossible(e.criterion.information)
resolvelib.resolvers.ResolutionImpossible: [RequirementInformation(requirement=<Requirement('torch==2.0.0+cpu')>, parent=None)]
```

Observation:

After some debugging, I found that `packages_from_links()` in `utils_pypi.py` operates on packages which have the version encoded. In particular, the `+` is encoded as `%2B`. My guess is that the matching somehow then compares `+` with `%2B` and fails. I tried the following (workaround) `requirements.txt`, which can be successfully analyzed. I guess this proofs that it's an encoding bug:

```
--extra-index-url https://download.pytorch.org/whl/cpu
torch==2.0.0%2Bcpu
```

### PR DESCRIPTION
**Pull request by fviernau**
`Distribution.from_link()` derives the version string of a package from the given (percent encoded) `Link.url`. That derivation lacks the decoding, so the resulting version string may also contain percent encoded characters in which case the dependency resolution fails.

Fix the resolution by URL adding the missing unquoting.

Fixes #143.

### CODE REVIEW

**Comment by TG1999** [comment]
@fviernau thanks ++, please have a look on https://github.com/nexB/python-inspector#testing to regenerate the tests with updated data

**Comment by fviernau** [comment]
> @fviernau thanks ++, please have a look on https://github.com/nexB/python-inspector#testing to regenerate the tests with updated data

@TG1999 I've re-generated  the test data and found that there were further failing tests. So, I've decided 
to make a dedicated PR to fix the tests. Please see https://github.com/nexB/python-inspector/pull/145.

Note: This PR should be merged only after #145 has been merged.

**Comment by fviernau** [comment]
@TG1999 I guess the failing tests for macos are just flakyness. In particular since they succeeded just recently for https://github.com/nexB/python-inspector/pull/145. Shall we re-run these failed checks?

**Comment by TG1999** [comment]
@fviernau yes, makes sense

**Comment by TG1999** [comment]
@fviernau the new test added here for do inputs with percent-encoded resolution is failing, which is an addition from #145 can you please have a look at it?

**Comment by fviernau** [comment]
> @fviernau the new test added here for inputs with percent-encoded resolution is failing, which is an addition from #145 can you please have a look at it?

I've tried to investigate this, but unfortunately all the tests do succeed on my machine running Ubuntu.
In order to investigate the above failing macos tests, I wouldn't know where to start. Do you have some pointers for me?
(I do not have a machine running macos at hand)

**Comment by pombredanne** [comment]
@fviernau From what I can see at https://download.pytorch.org/whl/torch/  macOs does not have a version `2.0.0+cpu` for Python 3.10 on macOS, therefore the resolution is not possible and fails as it should IMHO. See https://github.com/nexB/python-inspector/issues/143#issuecomment-1733827713 for the details on `+`.


> The exact scheme chosen is largely modeled on the existing behavior of pkg_resources.parse_version and pkg_resources.parse_requirements, with the main distinction being that where pkg_resources currently always takes the suffix into account when comparing versions for exact matches, the PEP requires that the local version label of the candidate version be ignored when no local version label is present in the version specifier clause

I understand this that when you ask for `torch==2.0.0+cpu` you can only get this version and nothing else. And there is no version for macOS that satisfies this requirement. Hence you need to design the test differently IMHO.

@TG1999 unrelated we will need to update this repo to use the latest skeleton and drop EOL Python 3.7 and older OS versions 
See https://www.python.org/downloads/release/python-3717/

**Comment by pombredanne** [comment]
@fviernau strike this out:
```
2023-09-25T12:25:32.1001420Z     @pytest.mark.online
2023-09-25T12:25:32.1001760Z     def test_get_resolved_dependencies_with_url_encoded_char_plus():
2023-09-25T12:25:32.1001940Z         req = Requirement("torch==2.0.0+cpu")
2023-09-25T12:25:32.1002250Z         req.is_requirement_resolved = True
2023-09-25T12:25:32.1002420Z         _, plist = get_resolved_dependencies(
2023-09-25T12:25:32.1002570Z             requirements=[req],
2023-09-25T12:25:32.1002700Z             environment=Environment(
2023-09-25T12:25:32.1002820Z                 python_version="310",
2023-09-25T12:25:32.1003670Z                 operating_system="linux",
2023-09-25T12:25:32.1003790Z             ),
2023-09-25T12:25:32.1003910Z             repos=[
2023-09-25T12:25:32.1004070Z                 PypiSimpleRepository(index_url="https://download.pytorch.org/whl/cpu", credentials=None)
2023-09-25T12:25:32.1004240Z             ],
2023-09-25T12:25:32.1004340Z >           as_tree=False,
2023-09-25T12:25:32.1004500Z         )
2023-09-25T12:25:32.1004570Z 
```
does NOT resolve on macOS but on linux.

I would like to see the test run on all supported Python versions and not only on 3.7, as it could be an older bug in pip on macOS on these versions

**Comment by fviernau** [comment]
> Hence you need to design the test differently IMHO.

I've spent some time looking for a library using local version identifier (`+`), which is available for all platforms without success. Looking at [1] it's probably not even worth search pypi.org.

@pombredanne  - Is your suggestion to use a single library for the test which is available to all platforms to all supported python versions? (If so, would you know such a library?)

[1] https://github.com/pypi/warehouse/issues/7989#issuecomment-632108464

File: tests/data/azure-devops.req-38-expected.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
     "tool_homepageurl": "https://github.com/nexB/python-inspector",
     "tool_version": "0.9.8",
     "options": [
-      "--requirement /home/frank/sources/github/nexB/python-inspector/tests/data/azure-devops.req.txt",
```
[the remark this answers is not in the document]
Comment (pombredanne) [reply tests/data/azure-devops.req-38-expected.json:7]: I created https://github.com/nexB/python-inspector/issues/150 ... @fviernau do the test still pass if you keep the original path? This seems really unrelated.

File: src/python_inspector/utils_pypi.py [hunk 2]
```diff
@@ -675,7 +676,7 @@ def from_link(cls, link: Link):
         """
         requires_python = link.python_requires
         path_or_url = link.url
-        filename = os.path.basename(path_or_url.strip("/"))
+        filename = os.path.basename(unquote(path_or_url).strip("/"))
```
Comment (pombredanne) [inline src/python_inspector/utils_pypi.py:679]: Are you sure you want to unquote there and not after getting the basename, so this happens just on the filename and not the whole path?

What if cls.`from_filename(filename)` directly with an encoded filename? Could it be a better place then?

**Review by pombredanne — CHANGES_REQUESTED** [review]
This looks good overall ... a few minor nits before merging:

1. Can you avoid making changes to the hardcoded test paths? this is minor but this is also unrelated.

2. Can you add a few unit tests for the `from_link()` method which is the main one impacted? https://github.com/nexB/python-inspector/pull/144/files#diff-f69b55183727cdb44af1543579e8ad3953ef4934462f87f8f052c2b6fb64e7abR671

OR moving the decoding to the `from_filename` as suggested below?

**Comment by pombredanne** [comment]
@TG1999 re: https://github.com/nexB/python-inspector/pull/144#issuecomment-1733908420 we need to update to the latest skeleton and drop CI OS combos  no longer available

**Comment by pombredanne** [comment]
> > Hence you need to design the test differently IMHO.
> 
> I've spent some time looking for a library using local version identifier (`+`), which is available for all platforms without any success. Looking at [1] it's probably not even worth search pypi.org.
> 
> @pombredanne - Is your suggestion to use a single library for the test which is available to all platforms to all supported python versions? (If so, would you know such a library?)
> 
> ...or rather: How do you think the test should look like?
> 
> [1] [pypi/warehouse#7989 (comment)](https://github.com/pypi/warehouse/issues/7989#issuecomment-632108464)

There is no such thing on PyPI alright. Let's get the latest cleaner CI defeinition and we may just mark this test as failing on macOS to move on as this is mysterious otherwise.

File: src/python_inspector/utils_pypi.py (hunk 2 above)
Comment (fviernau) [reply src/python_inspector/utils_pypi.py:679]: Makes sense to me. I've changed it, hope it's in the right place (overload) of that function.

**Comment by fviernau** [comment]
@pombredanne I believe I've addressed all above points. The test is for now disabled on macOs.

**Comment by pombredanne** [comment]
@fviernau FYI, I pushed a merge of main and a few commits to add tests, and refactored the tests approach to avoid any user-specific paths.

**Review by pombredanne — APPROVED** [review]

**Pull request merged**